relay-sampling/src/evaluation.rs

```diff
@@ -95,139 +114,193 @@ pub fn merge_configs_and_match(
 
-            if matches {
-                if let Some(evaluator) = SamplingValueEvaluator::create(rule, now) {
-                    matched_rule_ids.push(rule.id);
+        if matches {
+            if let Some(value) = rule.sample_rate(now) {
+                matched_rule_ids.push(rule.id);
 
-                    if rule.ty == RuleType::Trace {
-                        if let Some(dsc) = dsc {
-                            seed = Some(dsc.trace_id);
-                        }
+                if rule.ty == RuleType::Trace {
+                    if let Some(dsc) = dsc {
+                        seed = Some(dsc.trace_id);
                     }
+                }
 
-                    let value = evaluator.evaluate(now);
-                    match rule.sampling_value {
-                        SamplingValue::Factor { .. } => accumulated_factors *= value,
-                        SamplingValue::SampleRate { .. } => {
-                            return Some(SamplingMatch {
-                                sample_rate: (value * accumulated_factors).clamp(0.0, 1.0),
-                                seed: match seed {
-                                    Some(seed) => seed,
-                                    // In case we are not able to generate a seed, we will return a no
-                                    // match.
-                                    None => return None,
-                                },
-                                matched_rule_ids: MatchedRuleIds(matched_rule_ids),
-                            });
-                        }
+                match rule.sampling_value {
+                    SamplingValue::Factor { .. } => accumulated_factors *= value,
+                    SamplingValue::SampleRate { .. } => {
+                        let sample_rate = {
+                            let base = (value * accumulated_factors).clamp(0.0, 1.0);
+                            if adjust_sample_rate {
+                                dsc.and_then(|d| d.sample_rate)
+                                    .map(|client_sample_rate| {
+                                        DynamicSamplingContext::adjust_sample_rate(
+                                            client_sample_rate,
+                                            base,
+                                        )
+                                    })
+                                    .unwrap_or(base)
+                            } else {
+                                base
+                            }
+                        };
+
+                        let Some(seed) = seed else {
+                            return SamplingResult::NoMatch;
+                        };
+
+                        let is_kept = sampling_match(sample_rate, seed);
+
+                        return SamplingResult::Match {
+                            sample_rate,
+                            seed,
+                            matched_rules: MatchedRuleIds(matched_rule_ids),
```

Comment:
thoughts on making a constructor where you pass in the sample_rate, seed, and matched rules, and it will generate the is_kept value? 

its what i did in `sampling_match_from_parts` that i used for testing: https://github.com/getsentry/relay/pull/2514/files#diff-266da88d9f1e67f709b8b4286a60a6bb63ac51cf982ba7c3c46af957a9c6a123R117